overlay: Fix automatic overlay closing on click when <select> is active.

pratik-pc opened the pull request:
When `select` dropdown is open, click events outside `<select>` is registered to common parent node of `<select>` and `<select>` element on which user click. This led to overlay exiting sometimes.

This commit fixes this behaviour by making sure we don't close the overlay if the click was on overlay-container or inside it.

Fixes: zulip#33464.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
**Before:**

[Screencast from 2025-07-07 00-59-56.webm](https://github.com/user-attachments/assets/e89a7650-5fb7-4906-9eeb-82d56725c528)

**After:**

[Screencast from 2025-07-07 00-59-12.webm](https://github.com/user-attachments/assets/d5f01c3b-8804-434d-bf6a-f2ead793a0fa)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/overlays.ts:
```diff
@@ -174,8 +174,12 @@ export function initialize(): void {
-        // until it is found.
-        if ($target.is(".exit, .exit-sign, .overlay-content, .exit span")) {
+        // until it is found. We don't want to exit if div.overlay is inside
+        // overlay-container or at the same node level.
```
I'm still not sure I understand why this is the right logic; can you explain more precisely what the scenarios are and how you're thinking about them?

pratik-pc replied:
The if-else statement there was to accommodate two different html template earlier, one with `.overlay-container` inside `.overlay-content` (channel settings template) and another where node had both `.overlay-content` & `.overlay-container` (personal settings template). For the node that has both `.overlay-container` and `.overlay-content`(possibly any clicks inside overlay-container), we should ensure that we don't exit the overlay on any click events, since the click occurred within the actual overlay area.

pratik-pc replied:
I guess it's better to just check if $target has `.overlay-container` rather than using .closest since there are only two cases(`.overlay-container` is either inside or at the same level)

sahil839 replied:
Would it be better if we first make the HTML structure same for all overlays? Having it different for one overlay leads to this confusing code and may result in some bug in future as well.

sahil839 replied:
I was thinking of changing the HTML structure for settings overlay instead if we can do that keeping the behavior same as the current behavior.

pratik-pc replied:
Changed html structure for settings overlay.

alya commented:
Are you updating the PR in response to @timabbott 's feedback? You can tag @sahil839 for a review whenever you're ready.

pratik-pc commented:
@sahil839 ready for review

sahil839 commented:
The HTML structure for settings overlay still does not matches the structure for stream settings overlay.

Also, the current structure is somewhat confusing like class names, where `.overlay-content` is a parent element of `overlay-container`. And due to `.overlay-content` being a flex element, for streams overlay to close on clicking outside it, we require `overlay-content` class in `if ($target.is(...` if below -
```js
$("body").on("click", "div.overlay, div.overlay .exit", (e) => {
    let $target = $(e.target);

    if (document.getSelection()?.type === "Range") {
        return;
    }

    // if the target is not the div.overlay element, search up the node tree
    // until it is found.
    if ($target.is(".exit, .exit-sign, .overlay-content, .exit span")) {
        $target = $target.closest("[data-overlay]");
    } else if (!$target.is("div.overlay")) {
        // not a valid click target then.
        return;
    }

    if ($target.data("noclose")) {
        // This overlay has been marked explicitly to not be closed.
        return;
    }

    const target_name = $target.attr("data-overlay")!;

    close_overlay(target_name);

    e.preventDefault();
    e.stopPropagation();
});
```

I don't think we require the `overlay-content` to be a flex element so we can potentially remove that if possible and also improve the handler above. @karlstolley any suggestions?

karlstolley commented:
Whew. Wholly agreed that tripping over our own feet on things like `.overlay-content` being a parent of `.overlay-container` is clearly adding to some needless complexity and confusion here.

I'd probably work to refine the logic for where clicks *generally* register for closing the overlay. I must admit that I am unable to replicate the original issue in `main` in either Firefox or Chrome on Mac, so I also have to wonder whether there isn't some browser- or platform-specific quirk in play here.

I'm interested in the sprit of the `if ($target.data("noclose")) {` statement, and wonder if a possible approach here is just to set the `noclose` data attribute, or something similar, when a select has been opened. The trouble is that there's no good event that I'm aware of for when a select has been closed.

sahil839 commented:
Yeah, not sure how we would set the `noclose` data attribute without having a event for select being closed.

I can reproduce this in Chrome on ubuntu.

I logged value of `$(e.target)` in the click handler I shared above both when a `select` element is opened and closed. When `select` is not opened, `e.target` is `div.simplebar-content-wrapper` element in the left sidebar and when `select` is opened `e.target` is `div#settings_page.overlay-content.overlay-container` element. Not sure why this is happening, but this is the reason I think that making the HTML structure better and consistent and updating the click handler would help fix this bug as a consequence .

pratik-pc committed:
overlay: Remove `overlay-content` dependency from channels overlay.

Earlier, `overlay-content` was used as a flex element to center the
`overlay-container`.

This commit removes `overlay-content` and centers the overlay-container
using margin with available space left over after height of overlay.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from draft overlay.

Earlier, overlay-content was used as flex element to position drafts
overlay.

This commit removes overlay-content and positions the overlay-container
using available space left over after defining height as margin.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from edit history overlay.

This commit removes `overlay-content` dependency to center
overlay-container and position it using margin.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from reminders overlay.

This commit removes `overlay-content` to position the overlay
container and centers the overlay using margin from available
space left after defining the height.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from scheduled overlay.

This commit removes overlay-content to position the scheduled
message overlay and uses margin with available viewport space.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from groups overlay.

This commit removes overlay-content to position overlay-container
and uses margin on available viewport space to center the container.

pratik-pc committed:
overlay: Fix automatic overlay closing on click when <select> is active.

When `<select>` dropdown is open, click events outside `<select>` is
registered to common parent node of <select> and element on which
user clicked. This leads to overlay exiting sometimes when click
was registered on .overlay-content.

This commit fixes this behaviour by making settings overlay html
structure match that of channel settings overlay and remove
overlay-content used for transition.

Fixes: zulip#33464.

pratik-pc committed:
overlay: Remove `overlay-content` dependency from about-zulip overlay.

Earlier, overlay-content was used to transition the overlay vertically
when opened.

This commit now targets the transition rule to the overlay-container
and removes overlay-content from the overlay.

pratik-pc committed:
overlay: Use overlay-container to apply all rules in info overlay.

Earlier, overlay-content had some rules which were applied to the
same element.

This commit removes overlay-content and uses overlay-container to
apply the same css rules.

pratik-pc committed:
css: Refactor common overlay-container rules.

This commit refactors common overlay-container rules for overlay
and also removes reference to overlay-content.

pratik-pc committed:
overlay: Remove overlay-content reference in close handler.

Earlier, overlay-content was used to check click target to potentially
close overlays for specific overlays.

Overlay-content has since been removed and we dont require to
check for this in click target.

This commit removes the reference for this.

pratik-pc commented:
Removed `overlay-content` and changed settings overlay structure to stream settings overlay.

pratik-pc commented:
@zulipbot add "mentor review"

@ligmitz up for your review

ligmitz commented:
@pratik-pc Can you please fix the conflicts on this one ? Also, I don't have much to offer here so I think you can proceed with next review step. Thanks !

sahil839 commented on web/styles/dark_theme.css:
```diff
@@ -172,7 +172,7 @@
     }
 
     & div.overlay .flex.overlay-content > .overlay-container,
-    #settings_page,
```
I guess this was removed by mistake, since this class is not used for streams UI and checking out to this commit breaks the design for settings overlay in dark theme.

sahil839 commented on web/templates/draft_table_body.hbs:
```diff
@@ -1,10 +1,20 @@
+                    <div class="overlay-keyboard-shortcuts">
+                        {{#tr}}
+                            To restore a draft, click on it or press <z-shortcut></z-shortcut>.
+                            {{#*inline "z-shortcut"}}{{popover_hotkey_hints "Enter"}}{{/inline}}
```
This is added two times which results in - 
<img width="1738" height="366" alt="image" src="https://github.com/user-attachments/assets/7387f04a-051d-4813-ae4d-8ec71a37183a" />

pratik-pc replied:
Made a mistake during rebase

sahil839 reviewed (commented):
I noticed a couple of bugs that I posted about in above comments.

Changes look good in general and I can also not reproduce the bug of overlay being closed when select element is opened, even though I can no longer reproduce that in main as well.

pratik-pc commented:
Update
- Fixed above bugs


> I can also not reproduce the bug of overlay being closed when select element is opened, even though I can no longer reproduce that in main as well.

4e94324 commit returns early in the click handler because of the `mouse_drag.is_drag(e)`. Not sure why this would trigger mouse drag though

pratik-pc commented:
Alright so I tested a few things and it looks like the browser when natively handling the select dropdown, it eats up the `mousedown` event outside the dropdown (when dropdown is visible). Tested this in a separate code playground and same is true for system wide zulip `select`'s. So the `mousedown` handler in `mouse_drag` doesn't catch it.

sahil839 commented:
Yes, I also asked some help from AI tools and they mention that it is browser behavior maybe due to adding another layer on top to render the dropdown, but I cannot find any reliable source for this.

I think these changes can still be helpful even though bug is already fixed due to some unexpected, can be said incorrect as the click is being considered a drag, behavior. But yeah we may need to update the commit messages accordingly. @karlstolley thoughts?

karlstolley commented:
I'd probably be in favor of pulling the commits that remove `.overlay-content` into their own PR (don't refer to them as a "dependency," though; that's confusing). That's useful work in its own right--though rather than continuing to plop new selectors in `dark_theme.css`, it would be terrific to see that box-shadow addressed through a variable somehow.

Given that the bug seems to be somehow resolved in `main`, I'm a little less certain of committing those parts until we have a better handle on what's happening.
